### PR TITLE
[SYCL] Implement basic reduction for parallel_for() accepting nd_range

### PR DESCRIPTION
**Pull request by v-klochkov**
This patch adds the algorithm that implements 1 reduction in parallel_for().
It handles all types and operations, including user's custom ones.
The more efficient variants are on the way.

What is NOT supported by this patch:
- parallel_for(range, ...) // i.e. simple range without work-group sizes
- parallel_for(nd_range, reduction1, reduction1, ...) // i.e. more than
  1 reductions in paralell_for
- USM
- vector reductions (dims > 1 & #elements > 1)
- HOST. The implementation used in this patch uses barrier(), which
  is not supported on HOST yet.

### CODE REVIEW

**Comment by v-klochkov** [comment]
Please see the current proposal for reduction feature here: https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/Reduction/Reduction.md

The first patch adding reduction.hpp file and reduction classes is here:
https://github.com/intel/llvm/pull/1585

File: sycl/include/CL/sycl/detail/cg.hpp [hunk 1]
```diff
@@ -46,7 +46,7 @@ class interop_handler {
 
 public:
   using QueueImplPtr = std::shared_ptr<detail::queue_impl>;
-  using ReqToMem = std::pair<detail::Requirement*, pi_mem>;
+  using ReqToMem = std::pair<detail::Requirement *, pi_mem>;
```
Comment (AlexeySachkov) [inline sycl/include/CL/sycl/detail/cg.hpp:49]: This seems to be an unrelated change

File: sycl/include/CL/sycl/handler.hpp [hunk 2]
```diff
@@ -190,14 +227,24 @@ class __SYCL_EXPORT handler {
     return LambdaName == KernelName;
   }
 
+  /// Saves the location of user's code passed in \param CodeLoc for future
+  /// usage in finalize() method.
+  void saveCodeLoc(detail::code_location CodeLoc) { MCodeLoc = CodeLoc; }
+
+  /// Stores the given \param Event to the \param Queue.
+  /// Even thought MQueue is a field of handler, the method addEvent() of
```
Comment (AlexeySachkov) [inline sycl/include/CL/sycl/handler.hpp:235]:
```suggestion
  /// Even though MQueue is a field of handler, the method addEvent() of
```

?

File: sycl/test/reduction/reduction_nd_conditional.cpp [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+      ; // do nothing.
+  }
+};
+
+template <typename T, int Dim, class BinaryOperation>
+class Known;
+template <typename T, int Dim, class BinaryOperation>
+class Unknown;
+
+template <typename T>
```
Comment (AlexeySachkov) [inline sycl/test/reduction/reduction_nd_conditional.cpp:48]: You could create a header file for definition of this class, since it is used in several tests

File: sycl/include/CL/sycl/handler.hpp [hunk 4]
```diff
@@ -729,6 +810,305 @@ class __SYCL_EXPORT handler {
...
+    auto LocalReds = Redu.getReadWriteLocalAcc(WGSize + InefficientCase, *this);
+
+    // Get read accessor to the buffer that was used as output
+    // in the previous kernel. After that create new output buffer if needed
+    // and get accessor to it (or use reduction's accessor if the kernel
+    // is the last one).
+    auto In = Redu.getReadAccToPreviousPartialReds(*this);
+    auto Out = Redu.getWriteAccForPartialReds(NWorkGroups, KernelRun, *this);
+
+    if (!InefficientCase) {
```
Comment (capatober) [inline sycl/include/CL/sycl/handler.hpp:954]: May `if(EfficientCase)` be better than double negation from your point of view?

File: sycl/include/CL/sycl/handler.hpp [hunk 5]
```diff
@@ -729,6 +810,305 @@ class __SYCL_EXPORT handler {
...
+      using AuxName = typename detail::get_reduction_aux_2nd_kernel_name_t<
+          KernelName, KernelType>::name;
+      auto ReduIdentity = Redu.getIdentity();
+      parallel_for<AuxName>(Range, [=](nd_item<Dims> NDIt) {
+        size_t WGSize = NDIt.get_local_range().size();
+        size_t LID = NDIt.get_local_linear_id();
+        size_t GID = NDIt.get_global_linear_id();
+        // Copy the element to local memory to prepare it for tree-reduction
+        LocalReds[LID] = (GID < NWorkItems) ? In[GID] : ReduIdentity;
+        LocalReds[WGSize] = ReduIdentity;
```
Comment (capatober) [inline sycl/include/CL/sycl/handler.hpp:996]: Will it be better to allow only one `NDIt` write to `WGSize` to not have multiple write operations from different items?

File: sycl/include/CL/sycl/handler.hpp [hunk 6]
```diff
@@ -729,6 +810,305 @@ class __SYCL_EXPORT handler {
...
+        for (size_t CurStep = PrevStep >> 1; CurStep > 0; CurStep >>= 1) {
+          if (LID < CurStep)
+            LocalReds[LID] = BOp(LocalReds[LID], LocalReds[LID + CurStep]);
+          else if (LID == CurStep && (PrevStep & 0x1))
+            LocalReds[WGSize] = BOp(LocalReds[WGSize], LocalReds[PrevStep - 1]);
+          NDIt.barrier();
+          PrevStep = CurStep;
+        }
+
+        // Compute the the partial sum/reduction for the work-group.
```
Comment (capatober) [inline sycl/include/CL/sycl/handler.hpp:1013]:
```suggestion
        // Compute the partial sum/reduction for the work-group.
```

File: sycl/include/CL/sycl/handler.hpp [hunk 7]
```diff
@@ -729,6 +810,305 @@ class __SYCL_EXPORT handler {
 #endif
   }
 
+  /// Implements a command group function that enqueues a kernel that calls
+  /// user's lambda function \param KernelFunc and does one iteration of
+  /// reduction of elements in each of work-groups.
+  /// This version uses tree-reduction algorithm to reduce elements in each
+  /// of work-groups. At the end of each work-groups the partial sum is written
```
Comment (Pennycook) [inline sycl/include/CL/sycl/handler.hpp:817]:
```suggestion
  /// of work-groups. At the end of each work-group the partial sum is written
```

File: sycl/include/CL/sycl/handler.hpp [hunk 8]
```diff
@@ -729,6 +810,305 @@ class __SYCL_EXPORT handler {
...
+  void reduAuxCGFunc(const nd_range<Dims> &Range, size_t NWorkItems,
+                     size_t KernelRun, Reduction &Redu) {
+    size_t WGSize = Range.get_local_range().size();
+    size_t NWorkGroups = Range.get_group_range().size();
+
+    // The last work-group may be not fully loaded with work, or the work group
+    // size may be not power of those. Those two cases considered inefficient
+    // as they require additional code and checks in the kernel.
+    bool IsUnderLoaded = NWorkGroups * WGSize != NWorkItems;
+    size_t InefficientCase = (IsUnderLoaded || (WGSize & (WGSize - 1))) ? 1 : 0;
```
Comment (Pennycook) [inline sycl/include/CL/sycl/handler.hpp:935]: It took me a while to work out why this is declared as a `size_t`.

Do you think it would be clearer if `InefficientCase` was a `bool`?  You could shift the logic of whether you need to add 1 or 0 to line 945.

File: sycl/include/CL/sycl/handler.hpp [hunk 9]
```diff
@@ -729,6 +810,305 @@ class __SYCL_EXPORT handler {
...
+
+    bool IsUpdateOfUserAcc =
+        Reduction::accessor_mode == access::mode::read_write &&
+        NWorkGroups == 1;
+
+    // Use local memory to reduce elements in work-groups into 0-th element.
+    // If WGSize is not power of two, then WGSize+1 elements are allocated.
+    // The additional last element is used to catch reduce elements that
```
Comment (Pennycook) [inline sycl/include/CL/sycl/handler.hpp:943]:
```suggestion
    // The additional last element is used to catch elements that
```

File: sycl/include/CL/sycl/handler.hpp [hunk 10]
```diff
@@ -729,6 +810,305 @@ class __SYCL_EXPORT handler {
...
+      // The last group may be not fully loaded. Still register it as a group.
+      if ((NWorkItems % WGSize) != 0)
+        ++NWorkGroups;
+      auto Range =
+          nd_range<1>(range<1>(WGSize * NWorkGroups), range<1>(WGSize));
+
+      handler AuxHandler(QueueCopy, MIsHost);
+      AuxHandler.saveCodeLoc(MCodeLoc);
+
+      // The last kernel DOES write to reductions's accessor.
```
Comment (Pennycook) [inline sycl/include/CL/sycl/handler.hpp:1099]:
```suggestion
      // The last kernel DOES write to reduction's accessor.
```

File: sycl/source/handler.cpp [hunk 11]
```diff
@@ -13,12 +13,24 @@
...
-event handler::finalize(const cl::sycl::detail::code_location &Payload) {
-  sycl::event EventRet;
+
+void handler::addEventToQueue(shared_ptr_class<detail::queue_impl> Queue,
+                              cl::sycl::event Event) {
+  Queue->addEvent(std::move(Event));
+}
+
+event handler::finalize() {
+  // This block of code is needed only to 5th/default reduction implementation.
```
Comment (Pennycook) [inline sycl/source/handler.cpp:28]: Assuming that what you've pushed here is the default implementation:
```suggestion
  // This block of code is needed only for default reduction implementation.
```

File: sycl/test/reduction/reduction_nd_s0_dw.cpp [hunk 12]
```diff
@@ -0,0 +1,149 @@
+// RUN: %clangxx -fsycl %s -o %t.out
+// RUNx: env SYCL_DEVICE_TYPE=HOST %t.out
+// RUN: %CPU_RUN_PLACEHOLDER %t.out
+// RUN: %GPU_RUN_PLACEHOLDER %t.out
+// RUN: %ACC_RUN_PLACEHOLDER %t.out
+//==----------------reduction_ctor.cpp - SYCL reduction basic test ---------==//
```
Comment (Pennycook) [inline sycl/test/reduction/reduction_nd_s0_dw.cpp:6]:
```suggestion
//==----------------reduction_nd_s0_dw.cpp - SYCL reduction basic test ---------==//
```

Although I think @bader said before that we can drop these licenses from tests, if we want to.

File: sycl/test/reduction/reduction_nd_conditional.cpp (hunk 3 above)
Comment (Pennycook) [reply sycl/test/reduction/reduction_nd_conditional.cpp:48]: Agreed.  I'd be tempted to rename it something like `CustomType` or `CustomVec` as well, to highlight why you're defining a new class instead of using `sycl::vec`.

File: sycl/include/CL/sycl/detail/cg.hpp (hunk 1 above)
Comment (v-klochkov) [reply sycl/include/CL/sycl/detail/cg.hpp:49]: It was done by clang-format

File: sycl/include/CL/sycl/handler.hpp (hunk 2 above)
Comment (v-klochkov) [reply sycl/include/CL/sycl/handler.hpp:235]: Fixed. Thank you.

File: sycl/include/CL/sycl/handler.hpp (hunk 6 above)
Comment (v-klochkov) [reply sycl/include/CL/sycl/handler.hpp:1013]: Fixed. Thank you.

File: sycl/include/CL/sycl/handler.hpp (hunk 7 above)
Comment (v-klochkov) [reply sycl/include/CL/sycl/handler.hpp:817]: Fixed. Thank you.

File: sycl/include/CL/sycl/handler.hpp (hunk 9 above)
Comment (v-klochkov) [reply sycl/include/CL/sycl/handler.hpp:943]: Fixed.

File: sycl/source/handler.cpp (hunk 11 above)
Comment (v-klochkov) [reply sycl/source/handler.cpp:28]: Yes. Fixed.

File: sycl/include/CL/sycl/handler.hpp (hunk 4 above)
Comment (v-klochkov) [reply sycl/include/CL/sycl/handler.hpp:954]: I thought about that too. If it causes questions, it makes sense to fix it. Thank you.

File: sycl/include/CL/sycl/handler.hpp (hunk 8 above)
Comment (v-klochkov) [reply sycl/include/CL/sycl/handler.hpp:935]: Yes, I did that. Thank you.

File: sycl/include/CL/sycl/handler.hpp (hunk 10 above)
Comment (v-klochkov) [reply sycl/include/CL/sycl/handler.hpp:1099]: Fixed.

File: sycl/test/reduction/reduction_nd_s0_dw.cpp (hunk 12 above)
Comment (v-klochkov) [reply sycl/test/reduction/reduction_nd_s0_dw.cpp:6]: That is something new (I see that new sycl/test/abi/* tests don't have it. I dropped licenses.

File: sycl/test/reduction/reduction_nd_conditional.cpp (hunk 3 above)
Comment (v-klochkov) [reply sycl/test/reduction/reduction_nd_conditional.cpp:48]: Ok, I created a new reduction_utils.hpp header and renamed the type. Thank you.

File: sycl/include/CL/sycl/handler.hpp (hunk 5 above)
Comment (v-klochkov) [reply sycl/include/CL/sycl/handler.hpp:996]: Excuse me, I do not understand what you suggest here. Please give more details.

**Comment by v-klochkov** [comment]
Thank you for quick response/review. 
I added the additional changes as a separate (5th) commit of this PR.
Please re-review.

**Comment by v-klochkov** [comment]
@Pennycook 
Ops, I missed one more comment. This one. It will take some time for me to add another test case.


> Could you please add a few tests using transparent functors as in https://github.com/intel/llvm/blob/sycl/sycl/test/group-algorithm/reduce.cpp#L75? I don't think this needs to be exhaustive, so just adding variants to the "Check with various operations" tests would be good.
> 
> Do you think it's clear what an "Aux" kernel is? It's clear to me, but may not be clear to readers less familiar with reductions. I don't want this point to block the merge, but thought I'd bring it up in case you had any ideas for alternative names.
>

**Comment by v-klochkov** [comment]
> @Pennycook
> Ops, I missed one more comment.

This time I added a test to check reductions using transparent operators. Please see the 6th commit.

Regarding 'Aux' and naming. I don't see very good alternatives right now. Perhaps the comments, I added before reduAuxCGFunc() and inside inside parallel_for() lowering, help understand what they do.

File: sycl/source/handler.cpp [hunk 13]
```diff
@@ -13,12 +13,24 @@
...
 namespace sycl {
-event handler::finalize(const cl::sycl::detail::code_location &Payload) {
-  sycl::event EventRet;
+
+void handler::addEventToQueue(shared_ptr_class<detail::queue_impl> Queue,
+                              cl::sycl::event Event) {
+  Queue->addEvent(std::move(Event));
+}
+
+event handler::finalize() {
```
Comment (tovinkere) [inline sycl/source/handler.cpp:27]: Ordered_queue sends in the code location information through finalize as a parameter. This change will affect ordered_queue. Please ensure that ordered_queue is correct as well.

Comment (v-klochkov) [reply sycl/source/handler.cpp:27]: Would you please show the code that calls finalize() method and passes code_location to it?
I grepped all files in SYCL folder and did not find such calls.
The only finalize(code_loc) call was in queue_impl::submit_impl(), which is used by ordered_queue.

But I fixed submit_impl, so no additional changes required.

File: sycl/include/CL/sycl/handler.hpp (hunk 5 above)
Comment (capatober) [reply sycl/include/CL/sycl/handler.hpp:998]: I mean that you have `WGSize` work items: `0,1,..., WGSize-1`. Each of work-items updates `LocalReds[WGSize]`. Would it better to have something like that `if(LID==0){LocalReds[WGSize] = ...}`. So, only one work item (not all) updates the value.

File: sycl/source/handler.cpp (hunk 13 above)
Comment (tovinkere) [reply sycl/source/handler.cpp:27]: @v-klochkov  I looked at it too and the share the queue_impl, so everything looks good.

File: sycl/include/CL/sycl/handler.hpp [hunk 14]
```diff
@@ -729,6 +810,307 @@ class __SYCL_EXPORT handler {
...
+    size_t WGSize = Range.get_local_range().size();
+    size_t NWorkItems = NWorkGroups;
+    size_t KernelRun = 1;
+    while (NWorkItems > 1) {
+      // Before creating another kernel, add the event from the previous kernel
+      // to queue.
+      addEventToQueue(QueueCopy, MLastEvent);
+
+      // TODO: here the work-group size is not limited by user's needs,
+      // the better strategy here is to make the work-group-size as big
```
Comment (romanovvlad) [inline sycl/include/CL/sycl/handler.hpp:1088]: Why this is the better strategy?

File: sycl/include/CL/sycl/handler.hpp (hunk 5 above)
Comment (v-klochkov) [reply sycl/include/CL/sycl/handler.hpp:998]: If you don't mind I'll try your suggestion in the next patch.
It is a tiny fix that should not hold the commit, because it will not change performance in any way.

Having additional conditional in the code is also bad for performance. (Hmm, both the conditional and that duplicated write could be eliminated by vectorizer that combines ops across work-items), 
I'll check the device code and do or not do the additional fix with the next patch which is currently blocked by this PR.

File: sycl/include/CL/sycl/handler.hpp [hunk 15]
```diff
@@ -729,6 +810,307 @@ class __SYCL_EXPORT handler {
...
+    //    necessary to reduce all partial sums into one final sum.
+
+    // 1. Call the kernel that includes user's lambda function.
+    // If this kernel is going to be now last one, i.e. it does not write
+    // to user's accessor, then detach user's accessor from this kernel
+    // to make the dependencies between accessors and kernels more clean and
+    // correct.
+    if (NWorkGroups > 1)
+      dissociateWithHandler(Redu.MAcc);
+
```
Comment (romanovvlad) [inline sycl/include/CL/sycl/handler.hpp:1072]: Suggest creating separate handler even to run user's lambda. And set "command type" of this handler to something like a NOP or Aggregator so finalize does nothing for such a type.

File: sycl/include/CL/sycl/handler.hpp [hunk 16]
```diff
@@ -729,6 +810,307 @@ class __SYCL_EXPORT handler {
...
+    bool IsUpdateOfUserAcc =
+        Reduction::accessor_mode == access::mode::read_write &&
+        NWorkGroups == 1;
+
+    // Use local memory to reduce elements in work-groups into 0-th element.
+    // If WGSize is not power of two, then WGSize+1 elements are allocated.
+    // The additional last element is used to catch elements that could
+    // otherwise be lost in the tree-reduction algorithm.
+    size_t NumLocalElements = WGSize + (IsEfficientCase ? 0 : 1);
+    auto LocalReds = Redu.getReadWriteLocalAcc(NumLocalElements, *this);
```
[the remark this answers is not in the document]
Comment (v-klochkov) [reply sycl/include/CL/sycl/handler.hpp:841]: That is one of reasons why I have that func in reduction_impl class.
Because the buffer.hpp and accessor.hpp cannot be included in handler.hpp.
I'll change if you really insist/repeat the request, I would prefer this 'auto'.

File: sycl/include/CL/sycl/handler.hpp (hunk 14 above)
Comment (v-klochkov) [reply sycl/include/CL/sycl/handler.hpp:1088]: Because user may specify nd_range with small work-group because of some good reasons requirements for user's lambda, for example nd_range=(1M, 4), So, the main kernel must obey that and use WGSize=4, but it is obviously more efficient to use bigger WGSize for aux kernels that only reduce elements and do that much faster when WGSize is big (it also will require less calls of aux kernel, when the WGSize is bigger).

I'll update comment.

File: sycl/include/CL/sycl/handler.hpp (hunk 15 above)
Comment (v-klochkov) [reply sycl/include/CL/sycl/handler.hpp:1072]: What is the reasoning for that?
That approach seems more error prone. User may create many accessors, which get associated with 'this' handler and not with the new handler. New handler will not have any knowledge about those objects
User may do some additional calls for 'this' handler which will not have effect to new handler running user's code.

File: sycl/include/CL/sycl/handler.hpp [hunk 17]
```diff
@@ -729,6 +810,307 @@ class __SYCL_EXPORT handler {
...
+  /// Implements a command group function that enqueues a kernel that calls
+  /// user's lambda function \param KernelFunc and does one iteration of
+  /// reduction of elements in each of work-groups.
+  /// This version uses tree-reduction algorithm to reduce elements in each
+  /// of work-groups. At the end of each work-group the partial sum is written
+  /// to a global buffer.
+  ///
+  /// Briefly: user's lambda, tree-reduction, CUSTOM types/ops.
+  template <typename KernelName, typename KernelType, int Dims, class Reduction>
+  void reduCGFunc(KernelType KernelFunc, const nd_range<Dims> &Range,
```
[the remark this answers is not in the document]
Comment (v-klochkov) [reply sycl/include/CL/sycl/handler.hpp:822]: I'll try that.

File: sycl/include/CL/sycl/handler.hpp [hunk 18]
```diff
@@ -729,6 +810,307 @@ class __SYCL_EXPORT handler {
...
+
+    // 2. Run the additional aux kernel as many times as needed to reduce
+    // all partial sums into one scalar.
+    size_t WGSize = Range.get_local_range().size();
+    size_t NWorkItems = NWorkGroups;
+    size_t KernelRun = 1;
+    while (NWorkItems > 1) {
+      // Before creating another kernel, add the event from the previous kernel
+      // to queue.
+      addEventToQueue(QueueCopy, MLastEvent);
```
[the remark this answers is not in the document]
Comment (v-klochkov) [reply sycl/include/CL/sycl/handler.hpp:1085]: It is done to mimic the dependencies chain that is created for normal SYCL code without reductions. I.e.:
Q.submit( { user's CGFunc and lambda func})
while (NWorkGroups > 1)
  Q.submit( { Aux function to reduce elements; })

File: sycl/include/CL/sycl/handler.hpp [hunk 19]
```diff
@@ -729,6 +810,307 @@ class __SYCL_EXPORT handler {
...
+    //    If (N2 == 1) then the partial sum is written to user's accessor.
+    //    Otherwise, a new global buffer is created and partial sums are written
+    //    to it.
+    // 2) Call an aux kernel (if necessary, i.e. if N2 > 1) as many times as
+    //    necessary to reduce all partial sums into one final sum.
+
+    // 1. Call the kernel that includes user's lambda function.
+    // If this kernel is going to be now last one, i.e. it does not write
+    // to user's accessor, then detach user's accessor from this kernel
+    // to make the dependencies between accessors and kernels more clean and
```
[the remark this answers is not in the document]
Comment (v-klochkov) [reply sycl/include/CL/sycl/handler.hpp:1068]: There will be redundant dependencies. Probably, nothing incorrect.
Let suppose there are
1) initialization of user's reduction accessor
2) main CGFunc (with user's lambda and 1 iteration to reduce elements)
3) Aux kernel 1
4) Aux kernel 2
5) Final Aux kernel 3

If do not detach reduction accessor:                              If detach reduction accessor:
```
(1)                                                                (2) can be started even before (1)
(2) depends on (1)                                       (3) depends on (2)
(3) depends on (2)                                       (4) depends on (3)
(4) depends on (3)                                       (1) may init reduction's buffer here 
(5) depends on (4) and (1)                           (5) depends on (5) and (1)
```

So, do you think I need just to remove dissociateWithHandler?

File: sycl/include/CL/sycl/handler.hpp [hunk 20]
```diff
@@ -729,6 +810,307 @@ class __SYCL_EXPORT handler {
...
+    // correct.
+    if (NWorkGroups > 1)
```
[the remark this answers is not in the document]
Comment (v-klochkov) [reply sycl/include/CL/sycl/handler.hpp:1070]: If the mainCGFunc is the last kernel, i.e. if nd_range was (N, N), then mainCGFunc and the kernel in it will write to user's accessor. 
Otherwise (if NWorkGroups > 1, i.e. number of partial sums > 1), there will be another kernel following to the main kernel. and thus main kernel does not write to user's reduction variable and the dependency to user's variable is redundant.

File: sycl/include/CL/sycl/handler.hpp [hunk 21]
```diff
@@ -729,6 +810,307 @@ class __SYCL_EXPORT handler {
...
+                  Reduction &Redu) {
+
+    size_t NWorkItems = Range.get_global_range().size();
+    size_t WGSize = Range.get_local_range().size();
+    size_t NWorkGroups = Range.get_group_range().size();
+
+    bool IsUnderLoaded = (NWorkGroups * WGSize - NWorkItems) != 0;
+    bool IsEfficientCase = !IsUnderLoaded && ((WGSize & (WGSize - 1)) == 0);
+
+    bool IsUpdateOfUserAcc =
```
[the remark this answers is not in the document]
Comment (v-klochkov) [reply sycl/include/CL/sycl/handler.hpp:832]: Please explain what it changes.

IsUpdateOfUserAcc check includes 'NWorkGroups == 1', which is not known statically, and if so,then how this case is different from all other temp variables such as IsEfficientCase, etc.
Do you ask using 'const TYPE Var = <something>;'  when 'Var' is initialized once and not changed after?

File: sycl/include/CL/sycl/handler.hpp (hunk 17 above)
[the remark this answers is not in the document]
Comment (v-klochkov) [reply sycl/include/CL/sycl/handler.hpp:822]: I'll move that to sycl/intel/reduction.hpp and make them separate routines. I don't see any value in adding them as static method to 'reducer' or 'reduction_impl' class.

File: sycl/include/CL/sycl/handler.hpp (hunk 5 above)
Comment (capatober) [reply sycl/include/CL/sycl/handler.hpp:998]: OK. Thanks for the response
